src/Adapter/Presenter/AbstractLazyArray.php

```diff
@@ -0,0 +1,314 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+namespace PrestaShop\PrestaShop\Adapter\Presenter;
+
+use Doctrine\Common\Util\Inflector;
+
+/**
+ * This class is useful to provide the same behaviour than an array, but which load the result of each key on demand
+ * (LazyLoading).
+ *
+ * Example:
+ *
+ * If your want to define the ['addresses'] array access in your lazyArray object, just define the public method
+ * getAddresses() and add the annotation arrayAccess to it. e.g:
+ *
+ *     @arrayAccess
+ *     @return array
+ *
+ *     public function getAddresses()
+ *
+ * The method name should always be the index name converted to camelCase and prefixed with get. e.g:
+ *
+ * ['add_to_cart'] => getAddToCart()
+ *
+ * You can also add an array with already defined key to the lazyArray be calling the appendArray function.
+ * e.g.: you have a $product array containing $product['id'] = 10; $product['url'] = 'foo';
+ *       If you call ->appendArray($product) on the lazyArray, it will define the key ['id'] and ['url'] as well
+ *       for the lazyArray.
+ * Note if the key already exists as a method, it will be skip. In our example, if getUrl() is defined with the
+ * annotation @arrayAccess, the $product['url'] = 'foo'; will be ignored
+ *
+ */
+abstract class AbstractLazyArray implements \Iterator, \ArrayAccess, \Countable, \JsonSerializable
+{
+    /**
+     * @var \ArrayObject
+     */
+    private $arrayAccessList;
+
+    /**
+     * @var \ArrayIterator
+     */
+    private $arrayAccessIterator;
+
+    /**
+     * @var array
+     */
+    private $methodCacheResults = array();
+
+    /**
+     * AbstractLazyArray constructor.
+     * @throws \ReflectionException
+     */
+    public function __construct()
+    {
+        $this->arrayAccessList = new \ArrayObject();
```

Comment:
Can you use namespace please, it's easier to maintain and see what class are using in this file :)